crates/bevy_app/src/lib.rs

```diff
@@ -22,45 +22,81 @@ pub mod prelude {
     #[doc(hidden)]
     pub use crate::AppTypeRegistry;
     #[doc(hidden)]
-    pub use crate::{
-        app::App, CoreStage, DynamicPlugin, Plugin, PluginGroup, StartupSchedule, StartupStage,
-    };
+    pub use crate::{app::App, CoreSet, DynamicPlugin, Plugin, PluginGroup, StartupSet};
 }
 
-use bevy_ecs::schedule::StageLabel;
+use bevy_ecs::{
+    schedule::{Schedule, ScheduleLabel, SystemSet},
+    system::Local,
+    world::World,
+};
 
-/// The names of the default [`App`] stages.
+/// The names of the default [`App`] schedules.
 ///
-/// The relative [`Stages`](bevy_ecs::schedule::Stage) are added by [`App::add_default_stages`].
-#[derive(Debug, Hash, PartialEq, Eq, Clone, StageLabel)]
-pub enum CoreStage {
-    /// The [`Stage`](bevy_ecs::schedule::Stage) that runs before all other app stages.
+/// The corresponding [`Schedule`](bevy_ecs::schedule::Schedule) objects are added by [`App::add_default_schedules`].
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CoreSchedule {
+    /// The schedule that runs once when the app starts.
+    Startup,
+    /// The schedule that contains the app logic that is evaluated each tick of [`App::update()`].
+    Main,
+    /// The schedule that controls which schedules run.
+    ///
+    /// This is typically created using the [`CoreSchedule::outer_schedule`] method,
+    /// and does not need to manipulated during ordinary use.
+    Outer,
+}
+
+impl CoreSchedule {
+    /// An exclusive system that controls which schedule should be running.
+    ///
+    /// [`CoreSchedule::Startup`] will run a single time, and then [`CoreSchedule::Main`] will run on every later update.
+    pub fn outer_loop(world: &mut World, run_at_least_once: Local<bool>) {
+        if !*run_at_least_once {
+            world.run_schedule(CoreSchedule::Main);
```

Comment:
should this be setting run_at_least_once to true?

Reply:
We can probably just run `CoreSchedule::Main` on the first update too.
```rust
if (!*startup_has_run) {
    world.run_schedule(CoreSchedule::Startup);
    *startup_has_run = true;
}

world.run_schedule(CoreSchedule::Main);
```

Reply:
> We can probably just run CoreSchedule::Main on the first update too.

We shouldn't do this, as it means that you can't use `App::update` effectively in testing scenarios to verify that startup was done correctly.

Reply:
Fixed now.